src/ray/util/logging_utils.cc

```diff
@@ -0,0 +1,121 @@
+// Copyright 2025 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/util/logging_utils.h"
+
+#include <cstring>
+#include <functional>
+#include <future>
+#include <mutex>
+#include <vector>
+
+#include "ray/util/compat.h"
+#include "ray/util/pipe_logger.h"
+#include "ray/util/util.h"
+
+#if defined(_WIN32)
+#include <io.h>
+#endif
+
+namespace ray {
+
+namespace {
+
+#if defined(__APPLE__) || defined(__linux__)
+int GetStdoutHandle() { return STDOUT_FILENO; }
+int GetStderrHandle() { return STDERR_FILENO; }
+#elif defined(_WIN32)
+HANDLE GetStdoutHandle() { return _fileno(stdout); }
+HANDLE GetStderrHandle() { return _fileno(stderr); }
+#endif
+
+struct StreamRedirector {
+  // Used to synchronize the logging completion.
+  std::function<void()> completion_callback;
+  RedirectionFileHandle redirection_handle;
+};
```

Comment:
Can we simplify it so that RedirectionFileHandle has a Close method (or destructor) that only returns when everything is flushed to disk.

Reply:
It's possible if we change `RedirectionFileHandle`'s synchronization into cooperative style